src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs

```diff
@@ -202,6 +202,53 @@ protected override Expression VisitValues(ValuesExpression valuesExpression)
         return valuesExpression;
     }
 
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunctionExpression)
+    {
+        if (sqlFunctionExpression is { IsBuiltIn: true, Arguments: not null }
+            && string.Equals(sqlFunctionExpression.Name, "COALESCE", StringComparison.OrdinalIgnoreCase))
+        {
+            var type = sqlFunctionExpression.Type;
+            var typeMapping = sqlFunctionExpression.TypeMapping;
+            var defaultTypeMapping = _typeMappingSource.FindMapping(type);
+
+            // ISNULL always return a value having the same type as its first
+            // argument. Ideally we would convert the argument to have the
+            // desired type and type mapping, but currently EFCore has some
+            // trouble in computing types of non-homogeneous expressions
+            // (tracked in https://github.com/dotnet/efcore/issues/15586). To
+            // stay on the safe side we only use ISNULL if:
+            //  - all sub-expressions have the same type as the expression
+            //  - all sub-expressions have the same type mapping as the expression
+            //  - the expression is using the default type mapping (combined
+            //    with the two above, this implies that all of the expressions
+            //    are using the default type mapping of the type)
+            if (defaultTypeMapping == typeMapping
+                && sqlFunctionExpression.Arguments.All(a => a.Type == type && a.TypeMapping == typeMapping)) {
```

Comment:
Related to the above comment on the CLR type, I think we could compare the store type instead of referencing-comparing the type mapping instances... There's no guarantee in EF that there's only one type mapping instance for a given type mapping, and unless I'm mistaken the only thing that matters here for ISNULL is the actual store type. So comparing the store type should be safe and possibly allow for some more scenarios.

```suggestion
                && sqlFunctionExpression.Arguments.All(a => a.Type == type && a.TypeMapping?.StoreType == typeMapping?.StoreType)) {
```

Reply:
IIRC this would not recognize cases in which the mismatch is in one of the parameters (such as `varchar` vs `varchar(15)`).
Maybe the right approach here would be to make typemappings `IEquatable` and compare them

Reply:
> IIRC this would not recognize cases in which the mismatch is in one of the parameters (such as varchar vs varchar(15)).

I think StoreType must always be the full store type - including any facets. If that's not the case, we most probably have bugs somewhere else as well. Do you have a specific case in mind where this doesn't work?

> Maybe the right approach here would be to make typemappings IEquatable and compare them

I'm not sure... The point here is that in this specific context, the only thing we care about is the actual type in the database; in other words, various other details that distinguish type mappings from one another are irrelevant (value comparers, value converters...), since they affect EF client-side processing only. Assuming we have a 100% unambiguous representation of the database type - which is what StoreType is supposed to be - comparing those should be sufficient and allow for maximum coverage of this optimization, I think.

A general IEquatable would need to also compare e.g. value comparers (since it's possible that in other contexts that's relevant), and so would needlessly exclude some valid cases...